File: src-tauri/plugins/yames-mobile/android/src/main/java/YamesMobilePlugin.kt
package com.yames.metronome.mobile

import android.Manifest
import android.app.Activity
import android.content.ComponentCallbacks2
import android.content.Intent
import android.content.res.Configuration
import android.media.AudioAttributes
import android.media.AudioFocusRequest
import android.media.AudioManager
import android.net.Uri
import android.os.Build
import android.util.Log
import android.view.WindowManager
import android.webkit.WebView
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import app.tauri.PermissionState
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.Permission
import app.tauri.annotation.PermissionCallback
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Channel
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin

@InvokeArg
class BackgroundAudioArgs {
    var active: Boolean = false
    var title: String = ""
    var body: String = ""
    var stopLabel: String = ""
    var channelName: String = ""
}

@InvokeArg
class ToggleArgs {
    var active: Boolean = false
}

@InvokeArg
class OpenUrlArgs {
    var url: String = ""
}

@InvokeArg
class EventChannelArgs {
    lateinit var channel: Channel
}

/**
 * The Android half of the `yames-mobile` plugin.
 *
 * Five commands and three events, and every one of them is here because the
 * M00 spike found the app losing the beat without it. See the Rust half
 * (`src/lib.rs`) for the whole list and the reasoning.
 *
 * Nothing in this file runs on the audio thread, touches the engine, or knows
 * what tempo is playing. It keeps the process alive, it asks the system for
 * the right to make a sound, it keeps the screen on, and it answers Back.
 */
@TauriPlugin(
    permissions = [
        Permission(strings = [Manifest.permission.POST_NOTIFICATIONS], alias = "notifications"),
    ],
)
class YamesMobilePlugin(private val activity: Activity) : Plugin(activity) {
    private companion object {
        const val TAG = "YamesMobile"
    }

    private val audioManager: AudioManager
        get() = activity.getSystemService(Activity.AUDIO_SERVICE) as AudioManager

    private var focusRequest: AudioFocusRequest? = null
    private var holdingFocus = false

    /**
     * The one way back to the webview. Opened once at startup by
     * `set_event_channel`; every event below is a tagged message through it.
     */
    @Volatile
    private var events: Channel? = null

    /** Whether the app has something open that Back should close first. */
    @Volatile
    private var backIntercept = false

    /** Kept so the permission callback can finish the play the user asked for. */
    private var pendingStart: BackgroundAudioArgs? = null

    /**
     * The last system-bar insets measured, kept because they are measured
     * before the webview asks for them.
     *
     * The first layout pass happens while the page is still loading, so the
     * one event that matters most — the one that tells the app where the
     * status bar and the gesture bar are before it draws anything — would be
     * sent into a channel that does not exist yet. `setEventChannel` replays
     * this.
     */
    @Volatile
    private var lastInsets: JSObject? = null

    private val focusListener =
        AudioManager.OnAudioFocusChangeListener { change ->
            // Three kinds, not the four the brief sketched. A phone call does
            // not arrive as its own signal: the dialer takes transient focus
            // like any other app, and telling the two apart would mean asking
            // for READ_PHONE_STATE — a permission a metronome has no business
            // holding, to draw a distinction the frontend does not act on.
            val kind =
                when (change) {
                    AudioManager.AUDIOFOCUS_LOSS_TRANSIENT,
                    AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK,
                    -> "focus_lost"
                    AudioManager.AUDIOFOCUS_GAIN -> "focus_gained"
                    AudioManager.AUDIOFOCUS_LOSS -> "focus_lost_permanently"
                    else -> return@OnAudioFocusChangeListener
                }
            // Loud on purpose: audio focus is the one thing here that cannot
            // be reproduced from the app's own side, so when a phone does
            // something surprising this line is the whole diagnosis.
            Log.i(TAG, "audio focus change $change -> $kind")
            emit(JSObject().put("event", "audio_interrupted").put("kind", kind))
        }

    private val backCallback =
        object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                // M00: Back finished the activity mid-click and the process
                // was gone. A metronome on a music stand must not be one
                // stray edge-swipe from silence.
                if (backIntercept) {
                    emit(JSObject().put("event", "back_pressed"))
                } else {
                    activity.moveTaskToBack(true)
                }
            }
        }

    /**
     * What the system says when it wants memory back.
     *
     * The honest signal for "the band may go now". A decoded backing band is
     * tens of megabytes of samples, and the biggest background process is the
     * first one Android kills — so the app is told when its UI is gone
     * (`TRIM_MEMORY_UI_HIDDEN` and above) and when the phone is short while
     * the app is still on screen (`RUNNING_LOW` / `RUNNING_CRITICAL`). What to
     * do about it is the frontend's call, in `src/mobile/bandMemory.ts`;
     * nothing here knows what a jam is.
     */
    private val memoryCallbacks =
        object : ComponentCallbacks2 {
            override fun onTrimMemory(level: Int) {
                emit(JSObject().put("event", "memory_trim").put("level", level))
            }

            override fun onConfigurationChanged(newConfig: Configuration) {}

            @Deprecated("Superseded by onTrimMemory, and still called on older phones")
            override fun onLowMemory() {
                emit(
                    JSObject()
                        .put("event", "memory_trim")
                        .put("level", ComponentCallbacks2.TRIM_MEMORY_COMPLETE),
                )
            }
        }

    override fun load(webView: WebView) {
        activity.runOnUiThread {
            (activity as? AppCompatActivity)?.onBackPressedDispatcher?.addCallback(backCallback)
            watchWindowInsets()
        }
        activity.registerComponentCallbacks(memoryCallbacks)
        ClickService.onStopRequested = {
            emit(JSObject().put("event", "stop_requested"))
        }
    }

    /**
     * On screen / off screen, straight from the activity's own lifecycle.
     *
     * `onStop` and not `onPause`: a permission dialog over the app pauses it
     * without hiding it, and a band let go for a dialog that closes a second
     * later would be a re-decode nobody asked for. `onStop` is the activity
     * really being gone from the screen.
     */
    override fun onResume() {
        emit(JSObject().put("event", "app_visible").put("visible", true))
    }

    override fun onStop() {
        emit(JSObject().put("event", "app_visible").put("visible", false))
    }

    override fun onDestroy() {
        ClickService.onStopRequested = null
        activity.unregisterComponentCallbacks(memoryCallbacks)
        abandonFocus()
        stopService()
    }

    // -----------------------------------------------------------------------
    // Commands
    // -----------------------------------------------------------------------

    @Command
    fun setBackgroundAudio(invoke: Invoke) {
        val args = invoke.parseArgs(BackgroundAudioArgs::class.java)
        if (!args.active) {
            abandonFocus()
            stopService()
            invoke.resolve()
            return
        }

        // Ask for the notification permission the first time and only the
        // first time: once the user has answered, `getPermissionState` stops
        // returning PROMPT and playback proceeds either way. A denied prompt
        // costs the row in the shade, not the click.
        val needsPrompt =
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
                getPermissionState("notifications") == PermissionState.PROMPT
        if (needsPrompt) {
            pendingStart = args
            requestPermissionForAlias("notifications", invoke, "notificationPermissionResult")
            return
        }

        startPlayback(args)
        invoke.resolve()
    }

    @PermissionCallback
    fun notificationPermissionResult(invoke: Invoke) {
        pendingStart?.let { startPlayback(it) }
        pendingStart = null
        invoke.resolve()
    }

    @Command
    fun keepAwake(invoke: Invoke) {
        val args = invoke.parseArgs(ToggleArgs::class.java)
        activity.runOnUiThread {
            if (args.active) {
                activity.window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
            } else {
                activity.window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
            }
        }
        invoke.resolve()
    }

    @Command
    fun openUrl(invoke: Invoke) {
        val args = invoke.parseArgs(OpenUrlArgs::class.java)
        val uri = Uri.parse(args.url)
        // Only the web. Every caller is an About or support link, and an app
        // that hands any string it is given to `ACTION_VIEW` is an app that
        // can be talked into launching something else.
        if (uri.scheme != "https" && uri.scheme != "http") {
            invoke.reject("only http and https links can be opened")
            return
        }
        val intent = Intent(Intent.ACTION_VIEW, uri).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        try {
            activity.startActivity(intent)
            invoke.resolve()
        } catch (e: Exception) {
            invoke.reject("no app on this phone can open that link", e)
        }
    }

    @Command
    fun setBackIntercept(invoke: Invoke) {
        backIntercept = invoke.parseArgs(ToggleArgs::class.java).active
        invoke.resolve()
    }

    @Command
    fun setEventChannel(invoke: Invoke) {
        events = invoke.parseArgs(EventChannelArgs::class.java).channel
        // The bars were measured before this channel existed; the app cannot
        // lay itself out until it knows about them, so they go first.
        lastInsets?.let { emit(it) }
        activity.runOnUiThread { ViewCompat.requestApplyInsets(activity.window.decorView) }
        invoke.resolve()
    }

    // -----------------------------------------------------------------------
    // Internals
    // -----------------------------------------------------------------------

    private fun emit(payload: JSObject) {
        events?.send(payload)
    }

    /**
     * Tell the app where the status bar and the gesture bar actually are.
     *
     * The activity is edge-to-edge, so the webview is laid out behind both
     * system bars — and the CSS answer to that, `env(safe-area-inset-*)`, is
     * not the one it looks like. Those values describe the **display cutout**,
     * not the system bars: on the emulator `safe-area-inset-top` is the
     * camera cut-out's 128px and `safe-area-inset-bottom` is **0**, which is
     * why the gesture pill sat straight on top of the tab labels. Nothing the
     * web side can read would have told it otherwise.
     *
     * So the measurement comes from here, where the real numbers are, in CSS
     * pixels — the unit the stylesheet thinks in, which on a webview laid out
     * at `width=device-width, initial-scale=1` is the density-independent
     * pixel. `systemBars()` is the status and navigation bars;
     * `displayCutout()` is unioned in so a notch that is taller than the
     * status bar still gets cleared.
     *
     * Sent on every change rather than once: the keyboard, and a system-bar
     * mode the user changes in Android's own settings, both move these.
     */
    private fun watchWindowInsets() {
        val decor = activity.window.decorView
        ViewCompat.setOnApplyWindowInsetsListener(decor) { view, insets ->
            val bars =
                insets.getInsets(
                    WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout(),
                )
            val density = activity.resources.displayMetrics.density.toDouble()
            val payload = JSObject()
            payload.put("event", "window_insets")
            payload.put("top", bars.top / density)
            payload.put("right", bars.right / density)
            payload.put("bottom", bars.bottom / density)
            payload.put("left", bars.left / density)
            lastInsets = payload
            emit(payload)
            // Not consumed, and the decor view's own handling still runs:
            // this listener is here to read, not to change how the window
            // lays itself out.
            ViewCompat.onApplyWindowInsets(view, insets)
        }
        ViewCompat.requestApplyInsets(decor)
    }

    private fun startPlayback(args: BackgroundAudioArgs) {
        requestFocus()
        val intent =
            Intent(activity, ClickService::class.java)
                .setAction(ClickService.ACTION_START)
                .putExtra(ClickService.EXTRA_TITLE, args.title)
                .putExtra(ClickService.EXTRA_BODY, args.body)
                .putExtra(ClickService.EXTRA_STOP_LABEL, args.stopLabel)
                .putExtra(ClickService.EXTRA_CHANNEL_NAME, args.channelName)
        // A second start with new extras is how the notification's text is
        // updated when the tempo changes — the service is already running and
        // `startForeground` replaces the row rather than adding one.
        ContextCompat.startForegroundService(activity, intent)
    }

    private fun stopService() {
        activity.stopService(Intent(activity, ClickService::class.java))
    }

    private fun requestFocus() {
        if (holdingFocus) return
        val result =
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                val attributes =
                    AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_MEDIA)
                        .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                        .build()
                val request =
                    AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN)
                        .setAudioAttributes(attributes)
                        // Yames has one volume and no duck: a metronome at
                        // half volume under a navigation prompt is a metronome
                        // you cannot play to. Pause, then come back.
                        .setWillPauseWhenDucked(true)
                        .setOnAudioFocusChangeListener(focusListener)
                        .build()
                focusRequest = request
                audioManager.requestAudioFocus(request)
            } else {
                @Suppress("DEPRECATION")
                audioManager.requestAudioFocus(
                    focusListener,
                    AudioManager.STREAM_MUSIC,
                    AudioManager.AUDIOFOCUS_GAIN,
                )
            }
        holdingFocus = result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED
        Log.i(TAG, "audio focus requested, granted=$holdingFocus")
    }

    private fun abandonFocus() {
        if (!holdingFocus) return
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            focusRequest?.let { audioManager.abandonAudioFocusRequest(it) }
            focusRequest = null
        } else {
            @Suppress("DEPRECATION")
            audioManager.abandonAudioFocus(focusListener)
        }
        holdingFocus = false
        Log.i(TAG, "audio focus abandoned")
    }
}

File: src-tauri/plugins/yames-mobile/ios/Sources/YamesMobilePlugin.swift
import AVFoundation
import Tauri
import UIKit
import WebKit

/// The iPhone half of the `yames-mobile` plugin.
///
/// Same five commands and the same `audio_interrupted` event as the Android
/// half in `../../android` — the frontend (`src/mobile/native.ts`,
/// `useAndroidNative.ts`) is shared and does not branch on which phone it is
/// running on. What each command *means* is different, because the two systems
/// take a metronome away from you in different ways:
///
/// * Android needs a foreground service or the process is frozen minutes after
///   the screen goes off. iOS needs an `AVAudioSession` in the `.playback`
///   category plus the `audio` background mode, and then keeps the app running
///   for exactly as long as it is making sound. So `setBackgroundAudio` here is
///   "activate the audio session", and the notification text Android puts in
///   the shade has nowhere to go — iOS's equivalent row is the Now Playing
///   control, which needs remote-command handlers Yames has no buttons for.
/// * `.playback` is also what makes the click ignore the ring/silent switch.
///   A metronome that goes quiet because the phone is on silent is a metronome
///   you cannot trust on a music stand.
/// * There is no Back gesture on iOS, so `setBackIntercept` exists and does
///   nothing. It has to exist: the shared frontend calls it.
///
/// Nothing in this file touches the audio thread or knows what tempo is
/// playing. It asks the system for the right to make a sound, keeps the screen
/// awake, and reports when something took the sound away.
class YamesMobilePlugin: Plugin {
  /// The one way back to the webview, opened once at startup by
  /// `setEventChannel`. Every event below is a tagged message through it.
  private var events: Channel?

  /// Whether `AVAudioSession` is currently activated by us.
  private var sessionActive = false

  /// What the metronome wants: 48 kHz, and a buffer short enough that the
  /// click lands where the eye expects it. These are *preferred* values —
  /// iOS is free to give something else, and what it actually granted is
  /// logged so a device run can be read after the fact.
  private static let preferredSampleRate = 48000.0
  private static let preferredBufferDuration = 0.005

  override init() {
    super.init()
    // Before anything opens an output stream. The engine's cpal stream is
    // built during Tauri's setup, and an AudioUnit takes the session's format
    // as it finds it — so the category and the two preferred values are set
    // here, at plugin construction, rather than on the first press of play.
    configureSession()
    observeSystem()
  }

  override func load(webview: WKWebView) {
    // The category can be reset by the system across a media-services reset,
    // so re-assert it once the webview exists too. Setting it twice is free.
    configureSession()
  }

  deinit {
    NotificationCenter.default.removeObserver(self)
    deactivateSession()
  }

  // -----------------------------------------------------------------------
  // Commands
  // -----------------------------------------------------------------------

  @objc public func setBackgroundAudio(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(BackgroundAudioArgs.self)
    if args.active {
      activateSession()
    } else {
      deactivateSession()
    }
    invoke.resolve()
  }

  @objc public func keepAwake(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(ToggleArgs.self)
    DispatchQueue.main.async {
      UIApplication.shared.isIdleTimerDisabled = args.active
    }
    invoke.resolve()
  }

  @objc public func openUrl(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(OpenUrlArgs.self)
    guard let url = URL(string: args.url), let scheme = url.scheme?.lowercased(),
      scheme == "https" || scheme == "http"
    else {
      // Same rule as the Android half: every caller is an About or support
      // link, and an app that hands any string it is given to the system is
      // an app that can be talked into opening something else.
      invoke.reject("only http and https links can be opened")
      return
    }
    DispatchQueue.main.async {
      UIApplication.shared.open(url, options: [:], completionHandler: nil)
    }
    invoke.resolve()
  }

  /// Deliberately a no-op. iOS has no Back gesture; the command exists so the
  /// shared frontend does not have to know that.
  @objc public func setBackIntercept(_ invoke: Invoke) throws {
    _ = try invoke.parseArgs(ToggleArgs.self)
    invoke.resolve()
  }

  @objc public func setEventChannel(_ invoke: Invoke) throws {
    events = try invoke.parseArgs(EventChannelArgs.self).channel
    invoke.resolve()
  }

  // -----------------------------------------------------------------------
  // Internals
  // -----------------------------------------------------------------------

  private func emit(_ payload: JSObject) {
    events?.send(payload)
  }

  private func interrupted(_ kind: String) {
    // Loud on purpose, exactly as on Android: an interruption is the one
    // thing here that cannot be reproduced from the app's own side, so on a
    // device this line is the whole diagnosis.
    NSLog("[YamesMobile] audio interrupted -> %@", kind)
    emit(["event": "audio_interrupted", "kind": kind])
  }

  /// `.playback` / `.default`: the click is media, it ignores the silent
  /// switch, and it does not duck or mix with anything else by default.
  private func configureSession() {
    let session = AVAudioSession.sharedInstance()
    do {
      try session.setCategory(.playback, mode: .default, options: [])
      try session.setPreferredSampleRate(Self.preferredSampleRate)
      try session.setPreferredIOBufferDuration(Self.preferredBufferDuration)
    } catch {
      NSLog("[YamesMobile] could not configure the audio session: %@", "\(error)")
    }
  }

  private func activateSession() {
    guard !sessionActive else { return }
    let session = AVAudioSession.sharedInstance()
    do {
      try session.setActive(true, options: [])
      sessionActive = true
      // asked-vs-granted, the way the Android half prints its Oboe modes:
      // iOS honours neither value on demand, and on a device this is the
      // only place the real numbers appear.
      NSLog(
        "[YamesMobile] audio session active; asked %.0f Hz / %.1f ms, got %.0f Hz / %.2f ms",
        Self.preferredSampleRate, Self.preferredBufferDuration * 1000,
        session.sampleRate, session.ioBufferDuration * 1000)
    } catch {
      NSLog("[YamesMobile] could not activate the audio session: %@", "\(error)")
    }
  }

  private func deactivateSession() {
    guard sessionActive else { return }
    sessionActive = false
    do {
      // `.notifyOthersOnDeactivation` so whatever stood aside for the
      // metronome — a podcast, a backing track in another app — is told it
      // can come back.
      try AVAudioSession.sharedInstance().setActive(
        false, options: [.notifyOthersOnDeactivation])
    } catch {
      NSLog("[YamesMobile] could not deactivate the audio session: %@", "\(error)")
    }
  }

  private func observeSystem() {
    let centre = NotificationCenter.default
    centre.addObserver(
      self, selector: #selector(handleInterruption(_:)),
      name: AVAudioSession.interruptionNotification, object: nil)
    centre.addObserver(
      self, selector: #selector(handleRouteChange(_:)),
      name: AVAudioSession.routeChangeNotification, object: nil)
  }

  /// A call, a timer, Siri. The same two kinds the Android half reports from
  /// audio focus, so the frontend handler is shared and unchanged.
  @objc private func handleInterruption(_ note: Notification) {
    guard let raw = note.userInfo?[AVAudioSessionInterruptionTypeKey] as? UInt,
      let type = AVAudioSession.InterruptionType(rawValue: raw)
    else { return }

    switch type {
    case .began:
      // iOS has already stopped our audio by the time this arrives.
      sessionActive = false
      interrupted("focus_lost")
    case .ended:
      let options = AVAudioSession.InterruptionOptions(
        rawValue: note.userInfo?[AVAudioSessionInterruptionOptionKey] as? UInt ?? 0)
      guard options.contains(.shouldResume) else {
        // The system says do not resume — a call still in progress, or
        // another app that took over. Stay paused; the frontend keeps its
        // "the pause was ours" flag and the user's next press of play
        // clears it.
        NSLog("[YamesMobile] interruption ended without shouldResume; staying paused")
        return
      }
      activateSession()
      interrupted("focus_gained")
    @unknown default:
      break
    }
  }

  /// Headphones pulled out of the socket, or a Bluetooth device walking away.
  /// Pause — which is what every music app does, and what a player expects.
  @objc private func handleRouteChange(_ note: Notification) {
    guard let raw = note.userInfo?[AVAudioSessionRouteChangeReasonKey] as? UInt,
      let reason = AVAudioSession.RouteChangeReason(rawValue: raw)
    else { return }
    if reason == .oldDeviceUnavailable {
      interrupted("focus_lost")
    }
  }
}

// ---------------------------------------------------------------------------
// Argument types — the same JSON the Kotlin half parses, so the Rust half has
// one payload shape for both phones.
// ---------------------------------------------------------------------------

struct BackgroundAudioArgs: Decodable {
  let active: Bool
  /// Android puts these in the notification shade. iOS has nowhere to put
  /// them, and decodes them only so the two halves take the same payload.
  var title: String?
  var body: String?
  var stopLabel: String?
  var channelName: String?
}

struct ToggleArgs: Decodable {
  let active: Bool
}

struct OpenUrlArgs: Decodable {
  let url: String
}

struct EventChannelArgs: Decodable {
  let channel: Channel
}

@_cdecl("init_plugin_yames_mobile")
func initPlugin() -> Plugin {
  return YamesMobilePlugin()
}
